### ISSUE TITLE
Proposed changes to the config file format

### ISSUE THREAD

**Issue by hanzi** [frickelbruder/kickoff#28]
Before I dive into the implementation, I wanted to run this by you guys first. What do you think?
(Sorry for the long text. Below is an example if you just want to see the result.)
## The Problem

While trying to assemble a config file for one of my projects, I ran into a few inconsistencies and things that feel unnecessarily complicated. Here are my thoughts, in no particular order.
1. In a section, you cannot use a `target` node to overwrite the one in `defaults`. Instead, you have to call it `config` here _even though you’re overriding `target` with it_.
2. The name of the `uri` node is misleading. It contains the path, not a URI.
   URI: `https://www.example.com/foo/bar?baz=qux` (more specifically a URL, but whatever)
   Path: `/foo/bar`
3. Defining scheme, host and port separately seems rather useless. Not a big problem, but for any use case that I can imagine it should be enough to define a base (`https://www.example.com`) and a path (`/foo/bar`).
4. Since we already have to manually define names for our rules, there is actually no reason for them to HaveAWeirdNotationLikeThis. Without any infrastructural changes whatsoever, YAML would already support more human-readable names.
5. The syntax for setting rule options may be easy to parse, but is harder to write and to read. For example, `HttpRequestTime: [["maxTransferTime", 1500]]` looks a bit weird. Also, it mentions the same thing (download time) twice and uses different words (request time and transfer time.)
## My Proposal

Issues 1 and 2 are easily solved: Rename `config` to `target` and `uri` to `path`. Each section is always deep-merged with the `defaults` section so everything behaves as you would expect.

Issue 3 is a matter of opinion. If we choose to do this, my vote goes towards having `base` and `path` that simply get concatenated. So if you define `{ base: http://www.example.com/foo, path: /bar }`, the resulting URL would be `http://www.example.com/foo/bar`.
`path` defaults to `/`.
`base` would be a required setting for now, but later we could even add the option of setting it via command line (`bin/kickoff.php check foo.yml http://www.example.com`.)

So much for the easy part.
### Rule Sets

This references #5, #6 and #7.

The idea is to define groups of rules that can be applied in bulk, and that contain some sensible defaults.

Now the issues are about creating dedicated commands for those rule sets, but I don’t see any reason why we shouldn’t offer them in the config as well. So next to the `rules` node I would introduce a `rule sets` node that contains an array of strings:

``` yml
rule sets:
    - seo
    - security

rules:
    - …
```

This would require looking up the rule sets `seo` and `security` and simply merging them into the `rules` node internally.
### Natural Language Rule Names

Instead of `HttpRequestTime: [["maxTransferTime", 1500]]` you could just say `downloads faster than: 1500ms`.

`downloads faster than` would be the name of the rule and `1500ms` its option.

I was considering the following rules:
1. Rule names are case insensitive.
2. Punctuation is ignored, so `contains an <h1> tag` is interpreted as `contains an h1 tag`. Of course, this is limited by YAML itself. Can’t use colons or quotes too freely.
3. A rule can _optionally_ have a parameter, like shown above. It does not have to: `exists` would be a rule as well.
4. An option can either be a scalar value or an array or a hash. It’s the rule’s job to properly parse that.
5. Names and options should be chosen so that the resulting config is a reasonably readable list of English sentences without too much weird syntax. Again: limitations imposed by YAML.

See below for a full example of what I mean.
### Allow For A Simpler Version?

Another idea is that if neither `defaults` nor `sections` nodes exist, we just treat the config as if it were part of a `sections` block.

But I’m not sure whether that is actually helpful or just adding more fragmentation.

Or, we could drop the `sections` node entirely and just treat `defaults` as a special kind of section.
## Full Example

``` yml
defaults:
    target:
        host: https://www.google.com
        path: /
        headers:
            X-Foo: bar

    rules:
        - exists
        - is gzipped


sections:
    main page:
        target: { path: / }

        rule sets:
            - seo
            - security
            - performance

        rules:
            - downloads faster than: 1000ms
            - contains: "I’m Feeling Lucky"
            # The rules “exists” and “is gzipped” are included as well because
            # they are defined in the "defaults" section.


    non-existant page:
        target: { path: /404 }

        rules:
            - is missing


    all rules:
        target:
            host: http://www.example.com
            path: /foo/bar
            headers:
                X-Test: yes
                # disables the X-Foo header that has been set in the defaults
                X-Foo: ~

        rules:
            - enables xss protection
            - does not expose server software:
                tolerate server without version: true
            - downloads faster than: 1500ms
            - has expires header in the far future
            - contains etag
            - exists
            - is gzipped
            - disables content sniffing
            - is missing
            - responds 304 to if modified since
            - prevents framing
            - restricts framing to origin
            - sets httponly flag for cookie: session_id
            - sets secure flag for cookie: session_id
            - enables hsts:
                includes subdomains: true
            - title tag length is good:
                min: 10
                max: 50
            - meta description length is good
            - open graph properties are present
            - twitter properties are present
            - canonical link is present
            - has hreflang tag
            - has apple touch icons
            - does not have generator meta tag
            - contains: "foo bar"
            - contains an h1 tag
            - contains exactly one h1 tag
```

**Comment by frickelbruder** [comment]
hi @hanzi ,

I really like most of that. The config looks readable to me and if there are no drawbacks, I even like the better naming. I'll open an issue for each of the points I want to approve.

Points 1+2 are perfectly a valid and should be refactored.

As you stated, point 3 is a matter of opinion, and I think, it is not such a big thing to define them seperately and additionally we have those values easily handable in a variable without adding to much parse_url and if-isset-magic. So I would like to keep this as is.

Point 4: I like that.

Point 5: Yes it looks weird and was just to make things work. This could definitely be done better.

> Or, we could drop the sections node entirely and just treat defaults as a special kind of section.

No, there is also a Rules section, where custom rule classes can be placed.

> Rule Sets

I like it better, if those rule sets are kept in a separate command as already done with the seo check. I don't want this project to become overly complex with too many configuration possibilities. This makes the necessary learning curve too big and might overwhelm users. 

I hope I covered all of your points.

I'll open the issues later the day.

Thank you very much for your input.

Regards
Markus

**Comment by hanzi** [comment]
> No, there is also a Rules section, where custom rule classes can be placed.

Whoops, I totally forgot about that. Can we rename that node, though? I’d like to see a distinction between “Modify the list of available rules” and “Choose the rules that should apply”. At the moment, both are called `rules`, even though they’re different concerns. How about `rule definitions`?

> I like it better, if those rule sets are kept in a separate command as already done with the seo check. I don't want this project to become overly complex with too many configuration possibilities. This makes the necessary learning curve too big and might overwhelm users. 

I would argue that it actually lowers the learning curve since you can use one or two pre-defined rule collections instead of having to read through all the available rules and select the ones you think are necessary.

Instead of introducing a dedicated node for that I also thought about making it a “fake rule” like `- include rule set: seo`. But that felt like an ugly obfuscation and mix of concerns.

**Comment by frickelbruder** [comment]
Hi there,

in my opinion those presets are not really usable, since you normally have to configure those rules according to your needs. What we get with this option is another configuration element, which can be added to your configuration file, which adds some noise to your learning.
For me, those predefined rulesets are nothing more than a "quick check", just to see quickly, if there are any major issues. So a separate command will decouple "quick check" and "detailed configured checks". And as such, I'd rather like to keep this out for now.

As for the other points:
I opened a branch 2.0 and committed the changes, which were mostly taken from your suggestions. I refactored some minor things additionally. I will add some minor refactorings tomorrow or the day after and will most likely merge those changes into dev-master.

One bigger thing, I didn't do:
Your rules really look readable and all, but in the end, I sticked with the existing naming for several reasons:
- Most likely you will copy&paste them from somewhere, which is easier, if it is one string
- The prose seems sometimes to abstract too much: prevents framing might also be achieved by other ways
- The configuration is inconsistent, sometimes its 

    rule: value

On the other hand it can be 

    rule: 
        key: value
  
You are right, it might feel better for some rules to configure them "inline", but this makes it inconsistent. Maybe I will reconsider this in the future. But for now, I would really stick with the "weird naming" ;-)

Regards
Markus
